pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -102,6 +117,12 @@ public abstract class BasePartitionUpsertMetadataManager implements PartitionUps
   // Initialize with 1 pending operation to indicate the metadata manager can take more operations
   private int _numPendingOperations = 1;
   private boolean _closed;
+  // The lock and boolean flag below ensure only one thread can start preloading and preloading happens only once.
+  private final Lock _preloadLock = new ReentrantLock();
+  private volatile boolean _isPreloaded = false;
```

Comment:
Why do we need 2 flags? Can we initialize `_isPreloading` based on whether preload is enabled?

Reply:
simplified, by setting _isPreloading to true if preloading is enabled, otherwise false